clash-lib/src/Clash/Primitives/DSL.hs

```diff
@@ -293,6 +298,38 @@ unvec vName v@(ety -> Vector vSize eType) = do
   pure (map (TExpr eType . vIndex) [0..vSize-1])
 unvec _ e = error $ "unvec: cannot be called on non-vector: " <> show (ety e)
 
+-- | Deconstruct a 'Maybe' into its constructor 'Bit' and contents of its 'Just'
+-- field. Note that the contents might be undefined, if the constructor bit is
+-- set to 'Nothing'.
+deconstructMaybe ::
+  (HasCallStack, Backend backend) =>
+  -- | Maybe expression
+  TExpr ->
+  -- | Name hint for constructor bit, data
+  (Text, Text) ->
+  -- | Constructor represented as a Bit, contents of Just
+  State (BlockState backend) (TExpr, TExpr)
+deconstructMaybe e@TExpr{ety} (bitName, contentName)
+  | SP "GHC.Maybe.Maybe" [(_nothing, []),(_just, [aTy])] <- ety
```

Comment:
You might use [`showName` from template-haskell](https://hackage.haskell.org/package/template-haskell-2.18.0.0/docs/Language-Haskell-TH-Syntax.html#v:showName) to be resiliant against the definition of Maybe moving around.

```suggestion
  | SP tyNm [(_nothing, []),(_just, [aTy])] <- ety
  , tyNm == showName ''Maybe
```

Reply:
Done!